.github/workflows/main.yml

```diff
@@ -0,0 +1,131 @@
+name: CI
+
+on:
+  push:
+    branches:
+    - master
+  pull_request:
+    branches:
+    - master
+
+jobs:
+  lint:
+    name: Check lint
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
+      - name: Cache pip
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
+            ${{ runner.os }}-
+      - name: Install dependencies
+        run: |
+          set -xe
+          python -VV
+          python -m pip install --upgrade tox
+      - name: Check lint
+        run: tox -e check-lint
+
+  tests:
+    name: Python ${{ matrix.python-version }}
+    runs-on: ubuntu-latest
+    needs: lint
+
+    services:
+      postgres:
+        image: postgres:10
+        env:
+          POSTGRES_DB: procrastinate
+          POSTGRES_PASSWORD: password
+        # Set health checks to wait until postgres has started
+        options: >-
+          --health-cmd pg_isready
+          --health-interval 10s
+          --health-timeout 5s
+        ports:
+          - 5432:5432
+
+    strategy:
+      matrix:
+        python-version: [3.6, 3.7, 3.8]
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install dependencies
+        run: |
+          set -xe
+          python -VV
+          python -m pip install --upgrade tox tox-gh-actions
+      - name: Run tox targets for ${{ matrix.python-version }}
+        run: python -m tox
+        env:
+          PGDATABASE: procrastinate
+          PGHOST: localhost
+          PGUSER: postgres
+          PGPASSWORD: password
+          PYTEST_ADDOPTS: --cov-append
+      - name: Rename coverage report
+        run: mv .coverage .coverage-${{matrix.python-version}}
+      - name: Upload coverage artifact for ${{ matrix.python-version }}
+        uses: actions/upload-artifact@v2
+        with:
+          name: coverage-reports
+          path: .coverage-${{matrix.python-version}}
+
+  coverage:
+    name: Compute and upload coverage
+    runs-on: ubuntu-latest
+    needs: tests
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
+      - name: Install dependencies
+        run: |
+          set -xe
+          python -VV
+          python -m pip install --upgrade coverage
+      - name: Download coverages artifact
+        uses: actions/download-artifact@v2
+        with:
+          name: coverage-reports
+      - name: Combine coverage artifacts
+        run: python -m coverage combine .coverage-*
```

Comment:
If we combine the coverage runs, codecov looses a bit of information. I think itis already able to combine reports, but it's also able to say what coverage came from which report, which it won't in this configuration. Do we need to do this ? Wouldn't it be easier to upload coverage just after running the tests ?

To be honest though, I could accept if we lose codecov as long as we get a report on the coverage, but we've been bitten by this (nice and free !) service a few times.

Reply:
Because we check the python version in some parts of the code (`signals.py` if I remember correctly), running the coverage on the unitests of a single python version (python 3.8 for instance) is not sufficient to achieve 100% coverage. Additionaly, in the current CI implementation, the unittests are run in parallel on different workers, so we have to upload the `.coverage` files to a single worker once all the tests have completed. I could not find a better way to upload the coverage than combining the `.coverage` with `coverage combine`. If you know a better way, I will gladly replace what I have done ! :+1:

Reply:
Hm, I think my question lacked clarity. Let's chat about this in our native language :p

Reply:
conclusion: each stage will upload its own coverage data to codecov, and we'll have less code here :)